loopy/target/pyopencl.py

```diff
@@ -455,10 +455,47 @@ def alignment_requirement(self, type_decl):
     def get_kernel_executor_cache_key(self, queue, **kwargs):
         return queue.context
 
+    def preprocess_translation_unit_before_codegen(self, t_unit, epoint,
+                                                   passed_args_dict):
+
+        # {{{ ValueArgs -> GlobalArgs if passed as array shapes
+
+        from loopy.kernel.data import ValueArg, GlobalArg
+        import numpy as np
+        import pyopencl.array as cla
+
+        knl = t_unit[epoint]
+        new_args = []
+
+        for arg in knl.args:
+            if isinstance(arg, ValueArg):
+                if (arg.name in passed_args_dict
+                        and isinstance(passed_args_dict[arg.name], (cla.Array,
+                                                                    np.ndarray))
+                        and passed_args_dict[arg.name].shape == ()):
+                    arg = GlobalArg(name=arg.name, dtype=arg.dtype, shape=(),
+                                    is_output=False, is_input=True)
+
+            new_args.append(arg)
+
+        knl = knl.copy(args=new_args)
+
+        t_unit = t_unit.with_kernel(knl)
+
+        # }}}
+
+        return t_unit
+
     def get_kernel_executor(self, program, queue, **kwargs):
         from loopy.target.pyopencl_execution import PyOpenCLKernelExecutor
+
+        epoint = kwargs.pop("entrypoint")
+        program = self.preprocess_translation_unit_before_codegen(program,
```

Comment:
Are we inside the cached bit here? If not, this is going to totally murder our invocation overhead.

Reply:
Comes quite early, here: https://github.com/inducer/loopy/blob/e232bb1361479f2e6fd3bd7efd7f4ecce270947f/loopy/translation_unit.py#L339.

I *think* we are the cached bit.

Reply:
I agree.